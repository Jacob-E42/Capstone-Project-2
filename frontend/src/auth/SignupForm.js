import React, { useContext, useState } from "react";
import { Button, Form, FormGroup, Input, Label } from "reactstrap";
import { Link, useNavigate } from "react-router-dom";
import UserContext from "../UserContext";
import "./auth.css";
const SignupForm = () => {
	const { signup } = useContext(UserContext);
	const [formData, setFormData] = useState({
		firstName: "",
		lastName: "",
		email: "",
		password: "",
		interests: []
	});
	const [errors, setErrors] = useState({});
	const navigate = useNavigate();

	const handleChange = e => {
		const { name, value, type, checked } = e.target;

		if (type === "checkbox") {
			// Checkbox input
			setFormData(prevFormData => ({
				...prevFormData,
				interests: checked
					? [...prevFormData.interests, value]
					: prevFormData.interests.filter(item => item !== value)
			}));
		} else {
			setFormData(prevFormData => ({
				...prevFormData,
				[name]: value
			}));
		}
	};

	const handleSubmit = async e => {
		e.preventDefault();
		console.debug("handleSubmit");

		const validationErrors = validateForm(formData);
		if (Object.keys(validationErrors).length === 0) {
			const result = await signup(formData);
			console.log(result);
			if (result.success) {
				console.log("Form submitted:", formData);
				navigate("/profile");
			} else {
				setErrors(result.errors);
			}
		} else {
			setErrors(validationErrors);
		}
	};

	const validateForm = data => {
		const errors = {};

		if (!data.firstName.trim()) {
			errors.firstName = "First name is required";
		}

		if (!data.lastName.trim()) {
			errors.lastName = "Last name is required";
		}

		if (!data.email.trim()) {
			errors.email = "Email is required";
		} else if (!isValidEmail(data.email)) {
			errors.email = "Invalid email address";
		}

		if (!data.password.trim()) {
			errors.password = "Password is required";
		} else if (!isValidPassword(data.password)) {
			errors.password =
				"Password must be at least 8 characters long and contain at least one letter and one number";
		}

		console.log(data.interests);
		if (data.interests.length === 0) {
			errors.interests = "Select at least one interest";
		} else if (data.interests.length > 3) {
			errors.interests = "Select no more than three interests";
		}

		return errors;
	};

	const isValidEmail = email => {
		// Perform email validation here
		// You can use regular expressions or other validation libraries
		// Return true if the email is valid, false otherwise
		return /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/.test(email);
	};

	const isValidPassword = password => {
		// Perform password validation here
		// You can use regular expressions or other validation logic
		// Return true if the password is valid, false otherwise
		return /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d!@#$%^&*]{8,}$/.test(password);
	};

	return (
		<Form onSubmit={handleSubmit}>
			<FormGroup>
				<Label for="firstName">First Name</Label>
				<Input
					type="text"
					name="firstName"
					id="firstName"
					value={formData.firstName}
					onChange={handleChange}
					invalid={!!errors.firstName}
				/>
				{errors.firstName && <div className="error">{errors.firstName}</div>}
			</FormGroup>
			<FormGroup>
				<Label for="lastName">Last Name</Label>
				<Input
					type="text"
					name="lastName"
					id="lastName"
					value={formData.lastName}
					onChange={handleChange}
					invalid={!!errors.lastName}
				/>
				{errors.lastName && <div className="error">{errors.lastName}</div>}
			</FormGroup>
			<FormGroup>
				<Label for="email">Email</Label>
				<Input
					type="email"
					name="email"
					id="email"
					autoComplete="email"
					value={formData.email}
					onChange={handleChange}
					invalid={!!errors.email}
				/>
				{errors.email && <div className="error">{errors.email}</div>}
			</FormGroup>
			<FormGroup>
				<Label for="password">Password</Label>
				<Input
					type="password"
					name="password"
					id="password"
					autoComplete="current-password"
					value={formData.password}
					onChange={handleChange}
					invalid={!!errors.password}
				/>
				{errors.password && <div className="error">{errors.password}</div>}
			</FormGroup>
			<FormGroup>
				<Label for="interests">Interests</Label>
				<p>Please select one to three interests</p>
				<div>
					<Input
						type="checkbox"
						name="interests"
						id="interest1"
						value="interest1"
						onChange={handleChange}
					/>
					<Label
						for="interest1"
						className="ml-2">
						Interest 1
					</Label>
				</div>
				<div>
					<Input
						type="checkbox"
						name="interests"
						id="interest2"
						value="interest2"
						onChange={handleChange}
					/>
					<Label
						for="interest2"
						className="ml-2">
						Interest 2
					</Label>
				</div>
				<div>
					<Input
						type="checkbox"
						name="interests"
						id="interest3"
						value="interest3"
						onChange={handleChange}
					/>
					<Label
						for="interest3"
						className="ml-2">
						Interest 3
					</Label>
				</div>
				{errors.interests && <div className="error">{errors.interests}</div>}
			</FormGroup>
			<Button type="submit">Submit</Button>

			<Button
				type="button"
				tag={Link}
				to="/">
				Back
			</Button>
		</Form>
	);
};

export default SignupForm;
