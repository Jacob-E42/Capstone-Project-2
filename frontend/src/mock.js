import React, { useCallback, useState } from "react";
import UserContext from "./context_providers/UserContext";
import AlertContext from "./context_providers/AlertContext";
import Request from "./api";
import ApiRequest from "./api";
import ApiContext from "./context_providers/ApiContext";
import JournalContext from "./context_providers/JournalContext";

const demoUser = {
	id: 11,
	firstName: "testfirst",
	lastName: "testlast",
	email: "test@test.net",
	interests: ["baking", "jetskiing"]
};

const demoJournal = {
	id: 42,
	userId: 11,
	title: "Baking adventure",
	entryText:
		"Yesterday I tried baking for the first time like I've always wanted to. It was a complete disaster, but at least it was fun.",
	entryDate: "2023-07-24",
	emotions: null
};

const UserProvider = ({ children }) => {
	return <UserContext.Provider value={{ user: demoUser }}>{children}</UserContext.Provider>;
};

const AnonUserProvider = ({ children }) => {
	const [currentUser, setCurrentUser] = useState(null);

	const login = useCallback(
		async formData => {
			try {
				const request = new Request();
				const token = await request.login(formData);
				if (token) {
					setCurrentUser(demoUser);
					return { success: true };
				} else throw new Error("Log in token in missing");
			} catch (errors) {
				console.error("login failed", errors);
				return { success: false, errors };
			}
		},
		[setCurrentUser]
	);

	const signup = useCallback(
		async formData => {
			try {
				const request = new Request();
				const token = await request.signup(formData);
				console.log(token);
				if (token) setCurrentUser(demoUser);
				else throw new Error("missing sign in token");
				return { success: true };
			} catch (errors) {
				console.error("signup failed", errors);
				return { success: false, errors };
			}
		},
		[setCurrentUser]
	);

	return <UserContext.Provider value={{ currentUser, login, signup }}>{children}</UserContext.Provider>;
};

const AlertProvider = ({ children }) => {
	const [msg, setMsg] = useState("");
	const [color, setColor] = useState("primary");

	return <AlertContext.Provider value={{ msg, setMsg, color, setColor }}>{children}</AlertContext.Provider>;
};

const ApiProvider = ({ children }) => {
	const [request, setRequest] = useState(new ApiRequest());
	request.setToken(
		`eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MTEsImVtYWlsIjoidGVzdEB0ZXN0Lm5ldCIsImlhdCI6MTY5MjI5MDkxN30.8e12Cw-TMts0FmZALH61ch03kUMlW5QoCi60dlfhyWY`
	);

	return <ApiContext.Provider value={{ api: request }}>{children}</ApiContext.Provider>;
};

const JournalProvider = ({ children }) => {
	return <JournalContext.Provider value={{ journal: demoJournal }}>{children}</JournalContext.Provider>;
};

export { UserProvider, AnonUserProvider, AlertProvider, ApiProvider, JournalProvider };
