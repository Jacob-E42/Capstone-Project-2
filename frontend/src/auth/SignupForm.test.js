import React from "react";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { AlertProvider, AnonUserProvider } from "../mock";
import { MemoryRouter } from "react-router";
import SignupForm from "./SignupForm";

jest.mock("../api", () => ({
	login: jest.fn(() => Promise.resolve("mockToken"))
}));

test("SignupForm renders without crashing", () => {
	render(
		<MemoryRouter>
			<AnonUserProvider>
				<AlertProvider>
					<SignupForm />
				</AlertProvider>
			</AnonUserProvider>
		</MemoryRouter>
	);
});

test("SignupForm matches snapshot", () => {
	const { asFragment } = render(
		<MemoryRouter>
			<AnonUserProvider>
				<AlertProvider>
					<SignupForm />
				</AlertProvider>
			</AnonUserProvider>
		</MemoryRouter>
	);
	expect(asFragment()).toMatchSnapshot();
});

test("SignupForm displays username and password inputs", () => {
	render(
		<MemoryRouter>
			<AnonUserProvider>
				<AlertProvider>
					<SignupForm />
				</AlertProvider>
			</AnonUserProvider>
		</MemoryRouter>
	);
	expect(screen.getByPlaceholderText("Enter your email")).toBeInTheDocument();
	expect(screen.getByPlaceholderText("Enter a password, at least 8 characters")).toBeInTheDocument();
	expect(screen.getByPlaceholderText("Enter your first name")).toBeInTheDocument();
	expect(screen.getByPlaceholderText("Enter your last name")).toBeInTheDocument();
});
