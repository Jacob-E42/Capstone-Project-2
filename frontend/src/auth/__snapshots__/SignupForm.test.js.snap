// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm matches snapshot 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="firstName"
      >
        First Name
      </label>
      <input
        class="form-control"
        id="firstName"
        name="firstName"
        placeholder="Enter your first name"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="lastName"
      >
        Last Name
      </label>
      <input
        class="form-control"
        id="lastName"
        name="lastName"
        placeholder="Enter your last name"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        autocomplete="email"
        class="form-control"
        id="email"
        name="email"
        placeholder="Enter your email"
        required=""
        type="email"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="password"
      >
        Password
      </label>
      <input
        autocomplete="current-password"
        class="form-control"
        id="password"
        name="password"
        placeholder="Enter a password, at least 8 characters"
        required=""
        type="password"
        value=""
      />
    </div>
    <button
      class="btn btn-secondary"
      type="submit"
    >
      Submit
    </button>
    <a
      class="btn btn-secondary"
      href="/"
      type="button"
    >
      Back
    </a>
  </form>
</DocumentFragment>
`;
